lib/signal-list-view.js

```diff
@@ -1,132 +1,124 @@
-        value: "restart-kernel"
-      },
-      {
-        name: "Shut Down",
-        value: "shutdown-kernel"
-      }
+      { name: "Interrupt", value: "interrupt-kernel" },
+      { name: "Restart", value: "restart-kernel" },
+      { name: "Shut Down", value: "shutdown-kernel" }
     ];
 
     this.wsKernelCommands = [
-      {
-        name: "Rename session for",
-        value: "rename-kernel"
-      },
-      {
-        name: "Disconnect from",
-        value: "disconnect-kernel"
-      }
+      { name: "Rename session for", value: "rename-kernel" },
+      { name: "Disconnect from", value: "disconnect-kernel" }
+    ];
+
+    this.switchCommands = [
+      { name: "Switch to", value: "switch-kernel", payload: null }
     ];
 
     this.onConfirmed = null;
-    this.list.addClass("mark-active");
-    this.setItems([]);
+    this.selectListView = new SelectListView({
+      itemsClassList: ["mark-active"],
+      items: [],
+      filterKeyForItem: item => item.name,
+      elementForItem: item => {
+        const element = document.createElement("li");
+        element.textContent = item.name;
+        return element;
+      },
+      didConfirmSelection: item => {
+        log("Selected command:", item);
+        if (this.onConfirmed) this.onConfirmed(item);
+        this.cancel();
+      },
+      didCancelSelection: () => this.cancel(),
+      emptyMessage: "No running kernels for this file type."
+    });
   }
 
-  toggle() {
-    if (this.panel) {
-      this.cancel();
-    } else if (atom.workspace.getActiveTextEditor()) {
-      this.editor = atom.workspace.getActiveTextEditor();
-      this.attach();
-    }
+  _mapCommands(commands: Array<Object>, kernelSpec: Kernelspec) {
+    return commands.map(command => ({
+      name: `${command.name} ${kernelSpec.display_name} kernel`,
+      command: command.value,
+      payload: command.hasOwnProperty("payload") ? kernelSpec : null
+    }));
   }
 
-  attach() {
-    // get language from editor
-    this.storeFocusedElement();
-    if (!this.panel) {
-      this.panel = atom.workspace.addModalPanel({ item: this });
+  async toggle() {
+    if (this.panel != null) {
+      this.cancel();
     }
-    this.focusFilterEditor();
-    const grammar = store.grammar;
 
-    // disable all commands if no kernel is running
+    const grammar = store.grammar;
     const kernel = store.kernel;
-    if (!kernel) {
-      this.setItems([]);
+    let listItems = [];
+    if (!kernel) return;
+
+    if (kernel instanceof WSKernel)
+      listItems = this._mapCommands(
+        _.union(this.basicCommands, this.wsKernelCommands),
+        kernel.kernelSpec
+      );
+    else {
+      const otherKernels = await kernelManager.getAllKernelSpecsForGrammar(
+        grammar,
+        kernelSpecs => _.without(kernelSpecs, kernel.kernelSpec)
+      );
+
+      listItems = _.union(
+        this._mapCommands(this.basicCommands, kernel.kernelSpec),
+        _.flatMap(otherKernels, otherKernel =>
```

Comment:
I used `flatMap` because otherwise it was returning nested lists.

I'm not opposed to any other way of doing this if you have something in mind, `flatMap` was just the simplest for me at the time.